### DISCUSSION TITLE
Coverage only modified files

### DISCUSSION THREAD

**Discussion by brunoandradebr** [vitest-dev/vitest#4057] [Q&A]
Is it possible to run coverage only to modified files, like using lint-staged with regular test ? 

```javascript

...
// scripts:...
"test:staged": "vitest related --run", // <--- this works well,
"coverage:staged": "vitest run --coverage", // <-- for coverage is possible ? 

"lint-staged": {
    "src/**/*.{ts,tsx}": [
      "yarn lint",
      "yarn pretty",
      "yarn test:staged",
      "yarn coverage:staged"
    ]
  },

```

**Comment by AriPerkkio** [comment]
`vitest related --coverage` without enabling `coverage.all` should work.

**Reply by brunoandradebr** [reply]
thanks! I've just added --run "coverage:staged": "vitest related --coverage --run" to not be locked in lintstaged process.

**Comment by brunoandradebr** [comment]
Sorry, it`s not working. maybe I'm missing something

vite.config

```javascript

import react from '@vitejs/plugin-react'

import tsconfigPaths from 'vite-tsconfig-paths'
import { defineConfig } from 'vitest/config'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react(), tsconfigPaths()],
  test: {
    globals: true,
    environment: 'jsdom',
    coverage: {
      provider: 'v8',
      exclude: ['**/style*.ts'],
      all: false,
      statements: 100,
      lines: 100,
      branches: 100,
      functions: 100,
    },
    setupFiles: './src/test/setup.ts',
    passWithNoTests: true,
  },
})
```

package.json
```javascript

"test:staged": "vitest related --run",
"coverage:staged": "vitest related --coverage --run",

"lint-staged": {
    "src/**/*.{ts,tsx}": [
      "yarn lint",
      "yarn pretty",
      "yarn test:staged",
      "yarn coverage:staged"
    ]
 }
```

my commit pipeline:

![image](https://github.com/vitest-dev/vitest/assets/4542059/a7fe2e8b-b47a-490d-9fb7-2e9664f98eee)

Related is indicating only my modified component test but it's throwing error for global coverage not my modified coverage.

modified file coverage:

![image](https://github.com/vitest-dev/vitest/assets/4542059/49042637-fe36-49c5-a6e8-4ba89028cc62)
 
Am I missing something ?

**Reply by AriPerkkio** [reply]
> but it's throwing error for global coverage not my modified coverage

In the error message above it says the line coverage for modified files is only 82%, not 100%. The coverage report posted above is not showing what the total line coverage is. The total coverage is printed in the headers of report table.

If you want to calculate coverage per file, use [`coverage.perFile: true`](https://vitest.dev/config/#coverage-perfile).

**Comment by brunoandradebr** [comment]
This was my initial question. I only modified FormField files, not entire reported files. In my staged files was only FormField component and I wrote a full test coverage (as printed above) but coverage is throwing an error saying all coverage is 82%, and it's true BUT I would like to consider only for staged files coverage. Even if full coverage is not 100%. Is it possible ?

**Reply by AriPerkkio** [reply]
So `vitest related` will run test files which are using the source files that were changed. Currently the coverage report will include all source files that those test files are using.

Correct me if I'm not right, but here it would be ideal to show coverage for only the files that were changed, and ignore everything else? I don't think this is currently supported but could be implemented. Feel free to open new feature request for this.

**Reply by brunoandradebr** [reply]
What I'm trying to do is using lintstaged to get staged files from git and when I commit my staged files, my pipeline llint staged files, run pretty to only staged files, and test only staged files, and now I wanna run coverage only at those staged files, and only accept commit when staged files get coveraged above defined threshold.

Today when I run git commit -m "bla bla" my lint-staged pipeline run:

 1 - lint - if there are some lint error on staged files, git commit get canceled
 2 - pretty 
 3 - test - run test only on staged files, if some test is broken, commit get canceled

// what I'm trying to do 
 coverage - run coverage only on staged files, if coverage (only for staged files) goes under  threshold, commit get canceled

**Reply by AriPerkkio** [reply]
This requires that the staged files are passed as `coverage.include` argument. I think this could be done when using JS configuration of `lint-staged`: https://github.com/okonet/lint-staged#example-wrap-filenames-in-single-quotes-and-run-once-per-file

Something like:

```js
// .lintstagedrc.js
export default {
  "src/**/*.{ts,tsx}": (filenames) => `vitest related ${filenames.join(',')} --coverage.enabled --coverage.include {filenames.join(',')}`,
}
```

**Reply by vitaliyslion** [reply]
`... --coverage.include ${filenames.join(',')}` worked like a charm, thank you.
The only thing I additionaly added was `--watch=false`, as otherwise lintstaged gets stuck.
